File: saizad-mvvm/src/main/java/com/saizad/mvvm/service/BaseLocationWorker.java
package com.saizad.mvvm.service;

import android.content.Context;
import android.location.Location;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;
import androidx.work.Data;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.saizad.mvvm.SaizadLocation;

import sa.zad.easypermission.AppPermission;


abstract public class BaseLocationWorker extends Worker {

    public static final String WORK_COMPLETED = "work_completed";
    protected AppPermission locationPermission;
    protected final SaizadLocation saizadLocation;

    public BaseLocationWorker(@NonNull Context arg0, @NonNull WorkerParameters arg1, SaizadLocation saizadLocation, AppPermission locationPermission) {
        super(arg0, arg1);
        this.locationPermission = locationPermission;
        this.saizadLocation = saizadLocation;
    }

    @NonNull
    @Override
    public final Result doWork() {
        if (!locationPermission.isPermissionGranted(getApplicationContext())) {
            Log.d("BaseLocationWorker", "Permission not granted");
            return success(false);
        }
        try {
            final Location location = saizadLocation.getLocationObservable(new SaizadLocation.Builder().fastInterval(1).interval(1).build(), Looper.getMainLooper()).blockingFirst();
            Log.d("location_update", location.toString());
            return locationDoWork(location);
        }catch (Exception e){
            return success(false);
        }
    }

    @WorkerThread
    protected abstract Result locationDoWork(Location location);

    protected static Result success(boolean success){
        return Result.success(new Data.Builder().putBoolean(WORK_COMPLETED, success).build());
    }
}

File: saizad-mvvm/src/main/java/com/saizad/mvvm/service/BaseLocationWorker.kt
package com.saizad.mvvm.service

import android.content.Context
import android.location.Location
import android.os.Looper
import android.util.Log
import androidx.annotation.WorkerThread
import androidx.work.Data
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.saizad.mvvm.SaizadLocation
import sa.zad.easypermission.AppPermission

abstract class BaseLocationWorker(
    context: Context,
    workerParameters: WorkerParameters,
    protected val saizadLocation: SaizadLocation,
    protected var locationPermission: AppPermission
) : Worker(context, workerParameters) {

    override fun doWork(): Result {

        if (!locationPermission.isPermissionGranted(applicationContext)) {
            return success(false)
        }

        return try {
            val location = saizadLocation.getLocationObservable(
                SaizadLocation
                    .Builder()
                    .fastInterval(1)
                    .interval(1)
                    .build(),
                Looper.getMainLooper()
            ).blockingFirst()

            locationDoWork(location)
        } catch (e: Exception) {
            success(false)
        }
    }

    @WorkerThread
    protected abstract fun locationDoWork(location: Location): Result

    companion object {
        const val WORK_COMPLETED = "work_completed"
        protected fun success(success: Boolean): Result {
            return Result.success(
                Data.Builder().putBoolean(WORK_COMPLETED, success).build()
            )
        }
    }

}